File: src/Algorithm/Dynamic_Programing/Coin_Change/CoinChange.java
package Algorithm.Dynamic_Programing.Coin_Change;

public class CoinChange {
    public static void main(String[] args){
        int[] coins = {1,2,5};
        int amount = 11;
        System.out.println(coinChange(coins,amount));
    }

    private static int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount+1];
        for(int i=1;i<dp.length;i++){
            dp[i] = dp.length;
            for (int c:coins){
                if (i >= c){
                    dp[i] = Math.min(dp[i], dp[i-c]+1);
                }
            }
        }
        return dp[amount] == dp.length ? -1 : dp[amount];
    }
}

File: src/Algorithm/Dynamic_Programing/Coin_Change/coinChange.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int n)
    {
        // Initialize the DynamicProg vector
        vector<int> DynamicProg(n+1, 1e9);
        DynamicProg[0] = 0;
        for(int i = 1 ; i<=n ; i++)
            {
            for(int j = 0 ; j<coins.size(); j++)
            {
                if(i>=coins[j])
                    DynamicProg[i] = min( DynamicProg[i], DynamicProg[i-coins[j]]);
            }
            DynamicProg[i]++;
        }
        if(DynamicProg[n]>=1e9)return -1; //no possibilities of finding amount
        else return DynamicProg[n];
    }
};
